contracts/croncat-agents/src/contract.rs

```diff
@@ -1,31 +1,538 @@
+};
+use croncat_sdk_agents::msg::{AgentResponse, AgentTaskResponse, GetAgentIdsResponse};
+use croncat_sdk_agents::types::{Agent, AgentStatus};
+use croncat_sdk_core::types::Config;
 
-use crate::error::ContractError;
-use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
+pub(crate) const CONTRACT_NAME: &str = "crates.io:croncat-agents";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn instantiate(
-    mut _deps: DepsMut,
+    deps: DepsMut,
     _env: Env,
     _info: MessageInfo,
     _msg: InstantiateMsg,
 ) -> Result<Response, ContractError> {
-    todo!();
+    let config: Config = CONFIG.load(deps.storage)?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("paused", config.paused.to_string())
+        .add_attribute("owner_id", config.owner_id.to_string()))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetAgent {
+            account_id,
+            total_tasks,
+        } => to_binary(&query_get_agent(deps, env, account_id, total_tasks)?),
+        QueryMsg::GetAgentIds {} => to_binary(&query_get_agent_ids(deps)?),
+        QueryMsg::GetAgentTasks { account_id, slots } => {
+            to_binary(&query_get_agent_tasks(deps, env, account_id, slots)?)
+        }
+    }
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn execute(
-    _deps: DepsMut,
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::RegisterAgent {
+            payable_account_id,
+            cost,
+        } => register_agent(deps, info, env, payable_account_id, cost),
+        ExecuteMsg::UpdateAgent { payable_account_id } => {
+            update_agent(deps, info, env, payable_account_id)
+        }
+        ExecuteMsg::UnregisterAgent { from_behind } => {
+            unregister_agent(deps.storage, &info.sender, from_behind)
+        }
+        ExecuteMsg::WithdrawReward {} => withdraw_agent_balance(deps, &info.sender),
+        ExecuteMsg::CheckInAgent {} => accept_nomination_agent(deps, info, env),
+    }
+}
+
+fn query_get_agent(
+    deps: Deps,
+    env: Env,
+    account_id: String,
+    total_tasks: u64,
+) -> StdResult<Option<AgentResponse>> {
+    let account_id = deps.api.addr_validate(&account_id)?;
+    let agent = AGENTS.may_load(deps.storage, &account_id)?;
+    let a = if let Some(a) = agent {
+        a
+    } else {
+        return Ok(None);
+    };
+
+    let agent_status = get_agent_status(deps.storage, env, account_id, total_tasks)
+        // Return wrapped error if there was a problem
+        .map_err(|err| StdError::GenericErr {
+            msg: err.to_string(),
+        })?;
```

Comment:
I would like to have typed errors, so it's easier to test and debug!

Reply:
Multiple errors could be caught here, we don't know the exact error

Reply:
```suggestion
?
```
Do we need to wrap it into generic then? @deveusss